Support fixed database role db_accessadmin

tanscorpio7 opened the pull request:
### Description

Support fixed database role db_accessadmin in babelfish.
Member of db_accessadmin can do the following
1. CREATE SCHEMA
2. CREATE/DROP USER
3. ALTER USER
4. They always have connect permission

#### Engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/447
#### Extension PR: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2970

### Issues Resolved

[BABEL-5136]

### Test Scenarios Covered ###


### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

tanscorpio7 committed:
support db_accessadmin

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
empty commit to rerun actions

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
Merge branch 'server-database-roles-4-x' into BABEL-5136

Imran-imtiaz48 reviewed (commented):
### Review Summary

This update appears to refine the logic related to parsing and validating statements for different user roles (guest and owner) and the creation of database roles. It includes adjustments to the number of expected statements based on login status and the role of the user, along with improvements to the role-granting process.

### Code Review

1. **Statement Parsing Logic**:
   - The adjustment to `expected_stmt_num` based on whether the user is a guest or the owner is well thought out. It accounts for different conditions like the presence of logins and the role of the owner, adding clarity and control over how many statements are expected to be parsed.
   - One suggestion here is to add comments or documentation around the conditions that differentiate between when an owner is `sa` or not, to improve readability.

   ```c
   if (!owner_is_sa)
       expected_stmt_num = list_length(logins) > 0 ? 12 : 11;	
   else
       expected_stmt_num = list_length(logins) > 0 ? 11 : 10;
   ```

2. **Role Assignment**:
   - Granting the `dbo` role to the owner only if they are not `sa` adds an extra layer of security and role distinction, which is important for preventing unnecessary privilege escalation.
   - The use of `update_GrantRoleStmt` and `update_CreateRoleStmt` functions is logical and structured, ensuring that roles are correctly assigned and updated.

   ```c
   if (!owner_is_sa)
   {
       /* Grant dbo role to owner */
       stmt = parsetree_nth_stmt(res, i++);
       update_GrantRoleStmt(stmt, list_make1(make_accesspriv_node(dbo)),
                           list_make1(make_rolespec_node(owner)));
   }
   ```

   This logic is solid, but for added clarity, consider extracting the `if (!owner_is_sa)` block into a function to further modularize the code.

3. **Error Handling**:
   - The error reporting for mismatched statement counts is correctly implemented. However, a more informative error message would be helpful. Instead of a generic `ERROR`, provide details on the expected versus actual number of statements, which would make debugging easier.

   ```c
   if (list_length(res) != expected_stmt_num)
       ereport(ERROR, (errmsg("Expected %d statements, but found %d", expected_stmt_num, list_length(res))));
   ```

4. **General Improvements**:
   - It's great that the changes focus on proper role handling and enforcing statement expectations.
   - If possible, avoid magic numbers such as `12`, `11`, `10`, and `8`. It would enhance code maintainability if these values were replaced by well-named constants or enums to represent the different scenarios.

### Recommendations

- Add inline comments or documentation for more complex conditions like role differentiation (`owner_is_sa`) to make the code easier to follow.
- Consider modularizing the block that grants roles into its own function.
- Provide more detailed error messages in case of statement mismatch.
- Use named constants for the statement numbers to avoid potential confusion.

---

tanscorpio7 committed:
add basic tests

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix get_authid_user_ext_physical_name

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -875,45 +868,76 @@ is_user(Oid role_oid)
 		 * Only sysadmin can not be dropped. For the rest of the cases i.e., type
 		 * is "S" or "U" etc, we should drop the user
 		 */
```
Lets remove this comment as well.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -849,24 +849,17 @@ get_authid_login_ext_idx_oid(void)
 
 bool
-is_user(Oid role_oid)
+is_user(Oid role_oid, bool current_db_only)
```
Let's add comment for this function usage.

tanscorpio7 committed:
refactor

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -875,45 +868,76 @@ is_user(Oid role_oid)
+			}
+			else
+				is_user = true;
+		}
+
 		ReleaseSysCache(tuple);
+		pfree(type_str);
 	}
 
-	ReleaseSysCache(authtuple);
+	pfree(rolname);
 
 	return is_user;
 }
 
 bool
-is_role(Oid role_oid)
+is_role(Oid role_oid, bool current_db_only)
```
nit: Does it make sense to refactor is_user and is_role to avoid duplication?

tanscorpio7 replied:
Yes

tanscorpio7 replied:
Done

HarshLunagariya commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -875,45 +868,76 @@ is_user(Oid role_oid)
+
+				Assert(!isnull);
+
+				is_user = (strcmp(db_name_cstring, current_db_name) == 0);
```
Please make sure any of these db_names doesn't mismatch due to truncation of name(in of the values) and let's add testcase for it.

tanscorpio7 replied:
Just checked. db_names are truncated when stored in `sys.sys.babelfish_authid_user_ext`. So is `current_db_name`

tanscorpio7 replied:
Even sysdatabases stores mangled truncated name

HarshLunagariya commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -964,20 +988,32 @@ get_authid_user_ext_physical_name(const char *db_name, const char *login)
+		user_can_connect = DatumGetInt32(datum);
+
+		if (user_can_connect == 1 ||
+			(has_privs_of_role(get_role_oid(login, false), get_role_oid(db_accessadmin, false))))
```
Can you please explain this change? Let's add comment for clarification.
Does it imply that even if user is disabled then also it can be utilized if it is member of db_accessadmin?

tanscorpio7 replied:
Yes, db_accessadmin has connect privilege, so its member always have access, unless DENY is used which babelfish does not support.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -254,13 +271,17 @@ gen_dropdb_subcmds(const char *dbname, List *db_users)
 		}
 	}
+	/* Drop db_accessadmin*/
+	appendStringInfo(&query, "REVOKE CREATE ON DATABASE dummy FROM dummy; ");
+	appendStringInfo(&query, "DROP ROLE dummy; ");
```
Let's execute DROP OWNED BY db_accessadmin statement as well -- For seamless cleanup even if object ownership got assigned by mistake in interoperability or any miscellaneous cases.

tanscorpio7 replied:
Done

rishabhtanwar29 commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2348,13 +2348,16 @@ CREATE OR REPLACE PROCEDURE sys.sp_helpdbfixedrole("@rolename" sys.SYSNAME = NUL
 	BEGIN
 		SELECT CAST('db_owner' AS sys.SYSNAME) AS DbFixedRole, CAST('DB Owners' AS sys.nvarchar(70)) AS Description;
 	END
+	ELSE IF LOWER(RTRIM(@rolename)) IS NULL OR LOWER(RTRIM(@rolename)) = 'db_accessadmin'
```
No need to check nullability again in ELSE.

tanscorpio7 replied:
Changed the definition as a whole. For null input we should be returning all fixed db roles. Earlier implementation did not take this into account

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -849,24 +849,17 @@ get_authid_login_ext_idx_oid(void)
+	char    	*rolname;
 
-	authtuple = SearchSysCache1(AUTHOID, ObjectIdGetDatum(role_oid));
-	if (!HeapTupleIsValid(authtuple))
-		ereport(ERROR,
-				(errcode(ERRCODE_UNDEFINED_OBJECT),
-				 errmsg("role with OID %u does not exist", role_oid)));
-	rolname = ((Form_pg_authid) GETSTRUCT(authtuple))->rolname;
-	tuple = SearchSysCache1(AUTHIDUSEREXTROLENAME, NameGetDatum(&rolname));
+	rolname = GetUserNameFromId(role_oid, false);
+	tuple = SearchSysCache1(AUTHIDUSEREXTROLENAME, CStringGetDatum(rolname));
```
A lot of this logic is already there, I guess your target branch is not up to date.

tanscorpio7 replied:
I changed 
```
authtuple = SearchSysCache1(AUTHOID, ObjectIdGetDatum(role_oid));
if (!HeapTupleIsValid(authtuple))
ereport(ERROR,
		(errcode(ERRCODE_UNDEFINED_OBJECT),
			errmsg("role with OID %u does not exist", role_oid)));
rolname = ((Form_pg_authid) GETSTRUCT(authtuple))->rolname;
```
to
```
rolname = GetUserNameFromId(role_oid, false);
```
I don't think any other commit did this change.

tanscorpio7 replied:
Further changed this to make is_role & is_user into one function.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -1225,3 +1258,125 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
+	set_config_option("babelfishpg_tsql.sql_dialect", "tsql",
+						GUC_CONTEXT_CONFIG,
+						PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+	set_config_option("babelfishpg_tsql.migration_mode", physical_schema_name_exists("dbo") ? "single-db" : "multi-db",
+						GUC_CONTEXT_CONFIG,
+						PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
```
Let's consult @rishabhtanwar29 specifically for this change.

tanscorpio7 replied:
@rishabhtanwar29

HarshLunagariya commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -1225,3 +1258,125 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
+		{
+			ListCell *parsetree_item;
+
+			SetConfigOption("createrole_self_grant", "inherit", PGC_USERSET, PGC_S_OVERRIDE);
```
do we require inherit here?

tanscorpio7 replied:
Nope, but existing roles were created this way, so keeping it same for consistency.
I think we should have one single PR were we fix this for new and existing databases

HarshLunagariya commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -1225,3 +1258,125 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
+				wrapper->stmt_location = 0;
+
+				ProcessUtility(wrapper,
+							"(CREATE LOGICAL DATABASE )",
```
Let's change "CREATE LOGICAL DATABASE" to something meaning ful in this context.

rishabhtanwar29 replied:
+1, it is important to provide right query string so that logs contain meaning information.

tanscorpio7 replied:
Done

HarshLunagariya commented on contrib/babelfishpg_tsql/src/multidb.c:
```diff
@@ -249,6 +249,17 @@ rewrite_object_refs(Node *stmt)
 										"in Babelfish")));
 				}
+				else if (strcmp(role_name, DB_ACCESSADMIN) == 0)
+				{
+					if (has_privs_of_role(GetUserId(), get_role_oid(get_db_owner_name(db_name), false)))
+					{
+						/* only members of db_owner can alter drop members to fixed db roles */
+					}
+					else
+						ereport(ERROR,
+								(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+								 errmsg("Cannot alter the role '%s', because it does not exist or you do not have permission.", role_name)));
```
This doesn't look like a suitable place to do such checks -- currently we are doing it in rewriter. I would vote to do it in executor(just following the existing model).

tanscorpio7 replied:
Agreed

rishabhtanwar29 replied:
+1

tanscorpio7 replied:
Done, moved logic to processUtility and added a helper function `get_authid_user_ext_original_name`
grantee --> get original name from this (only in current db) --> if db_accessadmin then only allow db_owner to add/drop members.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3202,16 +3215,25 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+						Oid 		save_userid;
+						Oid 		db_owner = get_role_oid(get_db_owner_name(db_name), false);
+						Oid 		db_accessadmin = get_role_oid(get_db_accessadmin_role_name(db_name), false);
+						Oid 		user_name = get_role_oid(stmt->role->rolename, false);
```
replace it with user_id

tanscorpio7 replied:
Done

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3222,19 +3244,37 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 							{
-								if (strcmp(cur_user, dbo_name) != 0 &&
-									strcmp(cur_user, user_name) != 0)
+								if (is_member_of_db_owner || (isuser && is_member_of_db_accessadmin &&
+									!has_privs_of_role(user_name, db_owner)))
```
Please explain the second condition specially `&& !has_privs_of_role(user_name, db_owner))`

HarshLunagariya replied:
Isn't `|| user_name == GetUserId()` condition missing here?

tanscorpio7 replied:
> Please explain the second condition specially && !has_privs_of_role(user_name, db_owner))

db_accessadmin cannot rename a user who is a member of db_owner.

> Isn't || user_name == GetUserId() condition missing here?

A user can only change its default_schema. renaming itself or changing its own login mapping permission is not implicitly available.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3244,23 +3284,30 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 							{
-								if (strcmp(cur_user, dbo_name) != 0 &&
-									strcmp(cur_user, user_name) != 0)
+								if (is_member_of_db_owner)
```
Isn't || user_name == GetUserId() condition missing here?

tanscorpio7 replied:
same as [this reply](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2970#discussion_r1776727575)

HarshLunagariya replied:
As discussed offline, This should be fixed in older versions as well, right? If yes then we should open a separate PR for it.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3276,12 +3323,11 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 						PG_FINALLY();
 						{
-							SetCurrentRoleId(prev_current_user, true);
+							SetUserIdAndSecContext(save_userid, save_sec_context);
```
A bit off-topic from this change - Acc. to our offline discussion, This won't be required if we decrement the GUC level(of PG side), right? I think the same happens in `AtEOXact_GUC`. In PG code as well, they resetting the UserId explicitly. I assume UserId is not handled by GUC stack[Might not be correct].

tanscorpio7 replied:
Right, transaction abort/cleanup handles the clean up of this local CurrentUserId change. But I think this might be more safe, since we really do not want to leak this state.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3560,6 +3615,22 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+					else
+					{
+						owner_oid = rolspec ? get_rolespec_oid(rolspec, true) : InvalidOid;
+						if (OidIsValid(owner_oid) && OidIsValid(db_accessadmin) && !member_can_set_role(GetUserId(), owner_oid) &&
+							has_privs_of_role(GetUserId(), db_accessadmin) &&
+							(is_user(owner_oid, true) || is_role(owner_oid, true)))
```
What happens if owner_oid is invalid? It looks like it will allow the creation with current user as owner. Please correct me if it's not the case.

tanscorpio7 replied:
Do you mean the first check `OidIsValid(owner_oid)` is not enough ?

rishabhtanwar29 replied:
Let's add a comment regarding what this condition is checking.

tanscorpio7 replied:
Done

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -1225,3 +1258,125 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
+		db_owner = get_db_owner_name(dbname);
+		db_accessadmin = get_db_accessadmin_role_name(dbname);
+
+		initStringInfo(&query);
+
+		appendStringInfo(&query, "CREATE ROLE dummy ROLE dummy; ");
+		appendStringInfo(&query, "GRANT CREATE ON DATABASE dummy TO dummy; ");
+
+		parsetree_list = raw_parser(query.data, RAW_PARSE_DEFAULT);
```
This can be done only once outside of the while loop. Here we just need to update the rolename/dbname.

tanscorpio7 replied:
Done

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -1225,3 +1258,125 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
+*/
+PG_FUNCTION_INFO_V1(create_database_roles_for_all_dbs);
+Datum
+create_database_roles_for_all_dbs(PG_FUNCTION_ARGS)
```
Why not make the name specific to what role it is creating or do we intend to add more roles with this function in this release?

tanscorpio7 replied:
Yes, three more fixed db roles will use this.

shalinilohia50 replied:
Can we move the creation of db roles to another function like `create_db_roles_if_not_exists` mentioned here? https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3018/files 

Since our PRs have common changes, that would be better to add other db roles in the same function.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/multidb.c:
```diff
@@ -249,6 +249,17 @@ rewrite_object_refs(Node *stmt)
 				}
+				else if (strcmp(role_name, DB_ACCESSADMIN) == 0)
+				{
+					if (has_privs_of_role(GetUserId(), get_role_oid(get_db_owner_name(db_name), false)))
+					{
+						/* only members of db_owner can alter drop members to fixed db roles */
+					}
+					else
```
Why not write a standalone inverse `if` instead of this empty `if` and then else? Like:
```
if (!has_privs_of_role(GetUserId(), get_role_oid(get_db_owner_name(db_name), false)))
```
Anyways, why don't we do it it in processUtility instead of object_rewrite?

tanscorpio7 replied:
> Why not write a standalone inverse if instead of this empty if and then else? Like:

Readability, these are important permission checks and reading inverse condition felt a bit confusing.
Although I think this one might be better of written inverse.

> Anyways, why don't we do it it in processUtility instead of object_rewrite?

Doing it in object_rewrite meant we just need to compare the original name. Doing it in processUtility means we will need to fetch the original name from the physical name using babelfish catalog.
But since we have syscache now, I'll do that.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/multidb.c:
```diff
@@ -1414,6 +1428,21 @@ get_guest_role_name(const char *dbname)
+{
+	char	   *name = palloc0(MAX_BBF_NAMEDATALEND);
+
+	if (get_migration_mode() == SINGLE_DB && strcmp(dbname, "master") != 0
+	    && strcmp(dbname, "tempdb") != 0 && strcmp(dbname, "msdb") != 0)
```
Looks like we need to a utility function called `is_builtin_db(const char *dbname)` to check if it is one of master/tempdb/msdb.

tanscorpio7 replied:
Will change this to a macro

tanscorpio7 replied:
Done

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3560,6 +3615,22 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 					}
+					else
+					{
+						owner_oid = rolspec ? get_rolespec_oid(rolspec, true) : InvalidOid;
```
I think owner_oid should be `GetUserId()` when rolspec is invalid so that we create schema with current_user as owner.

tanscorpio7 replied:
We are trying to check if current user is trying to create schema for some one else.
If the current user itself is the owner, rolspec will be null and `CreateSchemaCommand` will handle it.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3560,6 +3615,22 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+					else
+					{
+						owner_oid = rolspec ? get_rolespec_oid(rolspec, true) : InvalidOid;
+						if (OidIsValid(owner_oid) && OidIsValid(db_accessadmin) && !member_can_set_role(GetUserId(), owner_oid) &&
```
Both of these owner and db_Accessadmin OIDs must be valid or otherwise we should throw error.

tanscorpio7 replied:
Owner oid will only be valid if `rolspec` exists. `rolspec` == NULL means current user should be owner.
for `db_Accessadmin` we can remove it based on how we moved forward with [this comment](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2970#discussion_r1776827788)

tanscorpio7 replied:
Actually I can refactor this a bit.
`else ---> else if (rolspec)` Then owner_oid will always be valid inside the block

tanscorpio7 replied:
Done

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3551,6 +3603,9 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 					PlannedStmt *wrapper;
 					RoleSpec *rolspec = create_schema->authrole;
+					Oid       owner_oid = InvalidOid;
+					Oid       db_accessadmin = get_role_oid(get_db_accessadmin_role_name(get_cur_db_name()), true);
```
Missing_ok should be false.

tanscorpio7 replied:
Problem here is during upgrade from 3.0 -> 3.1 we create guest schemas for all existing babelfish databases. 
During this CREATE SCHEMA operation db_accessadmin will not exists.
`create_guest_schema_for_all_dbs` <-- Problematic function name

Should I instead reuse this query string `(CREATE LOGICAL DATABASE )` in `create_guest_schema_for_all_dbs` ?
Then we can change missing okay to false.

tanscorpio7 replied:
Made some change and setting missing ok to false now

rishabhtanwar29 commented on test/JDBC/input/db_accessadmin-vu-cleanup.mix:
Let's add this test to upgrade schedule files.

tanscorpio7 committed:
refactor is_role() & is_user() to one function

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
address comments

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.3.0--4.4.0.sql:
```diff
@@ -1906,6 +1915,211 @@ CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'sp_statistics_inte
+LANGUAGE 'pltsql';
+GRANT EXECUTE ON PROCEDURE sys.sp_helpdbfixedrole TO PUBLIC;
+
+CREATE OR REPLACE PROCEDURE sys.sp_helpuser("@name_in_db" sys.SYSNAME = NULL) AS
```
There is already sys.sp_helpuser() in this upgrade script. I just figured it out while rebasing my PR over your branch.

tanscorpio7 replied:
Fixed

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
Merge remote-tracking branch 'upstream/BABEL_4_X_DEV' into BABEL-5136

tanscorpio7 committed:
empty commit to rerun actions

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
add test to upgrade

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
add test to upgrade

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
Merge remote-tracking branch 'origin/BABEL_4_X_DEV' into BABEL-5136

tanscorpio7 committed:
Merge remote-tracking branch 'origin/BABEL_4_X_DEV' into BABEL-5136

tanscorpio7 committed:
fix tests

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
empty

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
see dump restore actions logs

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
ssh into github actions host

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
remove temp changes and reset password for login in verify file

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
empty

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -848,74 +848,54 @@ get_authid_login_ext_idx_oid(void)
+ */
+
+const int
+is_database_principal(Oid role_oid, bool current_db_only)
```
Suggestion: The function no longer returns a boolean now so we can name it like `get_bbf_rolekind(Oid role_oid, const char *db_name)`. Note that passing `db_name` makes it more generic and usable.

tanscorpio7 replied:
Right, changed name to `get_db_principal_kind`. Did not want to use "role" in function name since it can be used for both pltsql database users and roles.
Also changes the bool flag to const char *db_name.

tanscorpio7 committed:
refactor

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

pranavJ23 committed:
fixing the issue of select cast with invalid number of function argument (#3005)

Cache lookup was failing when we ran cast query with extensions like PgAudit enabled.
The issue was because the arguments were not passed correctly into the cast functions. Added the handling cases as per the number of arguments(nargs) passed via cast query.

Issues Resolved : [BABEL-5272] [BABEL-4476]


Signed-off-by: Pranav Jain pranavjc@amazon.com

HarshLunagariya commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -254,15 +271,19 @@ gen_dropdb_subcmds(const char *dbname, List *db_users)
 		char	   *user_name = (char *) lfirst(elem);
 
-		if (strcmp(user_name, db_owner) != 0 && strcmp(user_name, dbo) != 0)
+		if (strcmp(user_name, db_owner) != 0 && strcmp(user_name, dbo) != 0 &&
+		    strcmp(user_name, db_accessadmin) != 0)
```
Can we use `IS_FIXED_DB_PRINCIPAL(role)` macro here?

tanscorpio7 replied:
No, the fixed role names are physical rolename here.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/session.c:
```diff
@@ -48,12 +48,23 @@ get_cur_db_id(void)
+ * returns current active babelfish db name
+ * Callers should NOT pfree the return value
+ */
+const char *
+get_current_db_name(void)
```
Better naming required IMO

tanscorpio7 replied:
Any suggestions ? I could not think of any better name. and it is returing const char so should be safe from pfree.

Yvinayak07 committed:
Fixed UDT handling for CASE statement when branch expression is of String Datatype. (#3009)

In CASE expression if we use UDT or sysname in any of the branch, there was no handling to find overall common type
and common typmod for CASE statement. For example, 

create type BABEL_CASE_EXPR_TEST_UDT from NVARCHAR(100);
GO

SELECT 
CASE 2
   WHEN 1 THEN 'abc'
   WHEN 2 THEN CAST('char' AS BABEL_CASE_EXPR_TEST_UDT)
END AS RESULT
GO

output:

Msg 8152, Level 16, State 1, Server BABELFISH, Line 1
value too long for type character varying(3)

Expected output:

nvarchar
char

Changes made to fix the issues - 

We added handling for UDT and sysname for CASE Expression. While finding common type of CASE Expression, we
checked if the following branch is a UDT or not, if yes then find the immediate base_type of UDT and find the TSQL
precedence according to the immediate base_type. If the branch if of type SYSNAME then we directly assign varchar as
datatype for that branch. While finding the common_typmod of CASE Expression, We checked if the branch is a UDT or
not, if yes then find the typmod of the base_type using getBaseTypeAndTypmod().If one of the branch is
of type SYSNAME then set typmod as SYSNAME_TYPMOD (i.e. 128).

Task: BABEL-5103, BABEL-4332
Signed-off-by: yashneet vinayak <yashneet@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3224,19 +3245,37 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 
 							if (strcmp(defel->defname, "default_schema") == 0)
 							{
-								if (strcmp(cur_user, dbo_name) != 0 &&
-									strcmp(cur_user, user_name) != 0)
+								if (is_member_of_db_owner || (isuser && is_member_of_db_accessadmin) ||
+									user_oid == GetUserId())
+								{
+									/*
+									 * members of db_owner can alter default schema for any role or user
+									 * members of db_accessadmin can alter default schema for any user
+									 */
+								}
+								else
+								{
 									ereport(ERROR,
 											(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
 											 errmsg("Current user does not have privileges to change schema")));
+								}
 							}
 							else if (strcmp(defel->defname, "rename") == 0)
 							{
-								if (strcmp(cur_user, dbo_name) != 0 &&
-									strcmp(cur_user, user_name) != 0)
+								if (is_member_of_db_owner || (isuser && is_member_of_db_accessadmin &&
+									!has_privs_of_role(user_oid, db_owner)))
```
As discussed offline, This should be fixed in older versions as well, right? If yes then we should open a separate PR for it.

tanscorpio7 replied:
I will file a separate JIRA for this but keeping it same as before in my PR feels wrong 😅

HarshLunagariya commented on test/JDBC/expected/single_db/db_accessadmin-vu-verify.out:
```diff
@@ -0,0 +1,337 @@
+
+
+-- db_accessadmin can not add/drop members to any other role
+ALTER ROLE db_accessadmin ADD MEMBER babel_5136_u1
+GO
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Cannot alter the role 'db_accessadmin', because it does not exist or you do not have permission.)~~
```
Please add testcases for
1. ALTER ROLE db_accessadmin ADD MEMBER [any other fixed role -- for now sysadmin & db_owner]
2. ALTER ROLE db_accessadmin DROP MEMBER
3. sp_addrolemember
4. sp_droprolemember

tanscorpio7 replied:
Added

HarshLunagariya commented on test/JDBC/expected/single_db/db_accessadmin-vu-verify.out:
```diff
@@ -0,0 +1,337 @@
+-- terminate-tsql-conn
+
+-- tsql user=babel_5136_l1 password=12345678 database=babel_5136
+-- Only db_owner members should be able to do this
+ALTER ROLE db_accessadmin ADD MEMBER babel_5136_db_accessadmin_user
+GO
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Cannot alter the role 'db_accessadmin', because it does not exist or you do not have permission.)~~
```
Let's add positive test case for this explicitly.

tanscorpio7 replied:
Already added, also added a comment above that line for easier test case review

ahmed-shameem committed:
Add '\' as ESCAPE character for sp_columns_100 for table_owner parameter (#3014)

There are few occurrences where we use procedure sp_columns_100 to fetch column metadata. The issue with this procedure is, when the table_owner (schema name) contains _ character, we may obtain some unexpected output . This is because in some places we are using using sp_columns_100 incorrectly with a backslash escaping underscores (e.g. “_“). Underscores are wildcards in sp_columns_100 (and in LIKE in general), but sp_columns_100 does not use a backslash as an escape character.
Till now, sp_columns_100 , was invoking a plpgsql function sp_column_100_internal which used PG’s LIKE operator for table_owner, which does use a backslash as the default escape character.

With BABEL-4953, we have updated the definition of sp_columns_100 where we removed redundant function call and implemented the procedure by using TSQL LIKE operator on column name, which will create some problems in other places.

The change in the new version is correct. However, since currently we rely on the incorrect functionality of the escape characters, we are making this temporary change for now. Once, we have fixed the dependencies on this procedure, we need to revert it.

Authored-by: Shameem Ahmed shmeeh@amazon.com
Signed-off-by: Shameem Ahmed shmeeh@amazon.com

HarshLunagariya commented on test/JDBC/expected/single_db/db_accessadmin-vu-verify.out:
```diff
@@ -0,0 +1,337 @@
+
+~~ERROR (Message: permission denied for database jdbc_testdb)~~
+
+CREATE DATABASE new_database
+GO
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Only one user database allowed under single-db mode. User database "babel_5136" already exists)~~
```
I think this is not intended. Please update.

tanscorpio7 replied:
Updated

HarshLunagariya commented on test/JDBC/expected/single_db/db_accessadmin-vu-verify.out:
```diff
@@ -0,0 +1,337 @@
+
+~~ERROR (Message: permission denied for schema dbo)~~
+
+CREATE PROCEDURE babel_5136_p1 AS SELECT 1
+GO
+~~ERROR (Code: 2714)~~
+
+~~ERROR (Message: Function 'babel_5136_p1' already exists with the same name)~~
+
+CREATE FUNCTION f1() RETURNS INT AS BEGIN RETURN 1 END
+GO
+~~ERROR (Code: 2714)~~
+
+~~ERROR (Message: Function 'f1' already exists with the same name)~~
+
+CREATE DATABASE new_database
+GO
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Only one user database allowed under single-db mode. User database "babel_5136" already exists)~~
```
This is not intended output. Please update.

tanscorpio7 replied:
Updated

HarshLunagariya commented on test/JDBC/expected/single_db/db_accessadmin-vu-verify.out:
```diff
@@ -0,0 +1,337 @@
+
+
+-- should be a member of db_accessadmin in babel_5136 database
+SELECT IS_ROLEMEMBER('db_accessadmin')
```
Let's test If dbo & db_owner is member of 'db_accessadmin' or not by default.

HarshLunagariya replied:
\+ Please add testcases for is_member as well.

tanscorpio7 replied:
Fixed `IS_ROLEMEMBER()` 
"dbo" should be a member of all fixed roles ONLY
No fixed db role can be a member of anyone else, so always zero.

HarshLunagariya commented on test/JDBC/expected/single_db/db_accessadmin-vu-verify.out:
```diff
@@ -0,0 +1,337 @@
+GO
+
+-- CREATE SCHEMA should be allowed db_accessadmin
+CREATE SCHEMA s1
+GO
+CREATE SCHEMA s2 AUTHORIZATION dbo
+GO
+CREATE SCHEMA s3 AUTHORIZATION babel_5136_u1
+GO
```
Let's add testcase for GRANT ON SCHEMA.

tanscorpio7 replied:
Done

HarshLunagariya commented on test/JDBC/expected/single_db/db_accessadmin-vu-verify.out:
```diff
@@ -0,0 +1,337 @@
+~~ERROR (Message: Only one user database allowed under single-db mode. User database "babel_5136" already exists)~~
+
+
+
```
Please make sure that
1. It is being tested in both the migration mode. 
2. we are tracking the testcase about the restrictions on the role that it should not be altered/dropped from either of endpoint as separate effort.
3. we test SSMS utility as well for visibility and usage of this new role via UI.
4. We are verifying that it should not be the owner of anything else apart from schema and schema-contained objects.

tanscorpio7 committed:
fix tests

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

Yvinayak07 committed:
Updating the upgrade test file after fixing the typmod for CASE expression (#3015)

Updating the upgrade test file after fixing the typmod for CASE expression. Need to change the 15_9 schedule file in
BABEL_4_X_DEV, as PR for BABEL_4_X_DEV is merged first, Which includes before files in 15_9 schedule file. But as
these changes need to be merged in BABEL_3_X_DEV, we need to add after files in place of before files as 15_9
corresponds to BABEL_3_X_DEV.

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

tanscorpio7 committed:
fix valgrind warning in pltsql_store_view_definition (#3019)

Reset static variable original_query_string back to NULL after use otherwise next queries which do not set original_query_string in ANLTR will read already freed memory.
---------
Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
refactor

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

robverschoor committed:
Support unary '+' operators for strings (#3013)


The T-SQL grammar allows an arbitrary number of unary + operators to precede an expression,
but PG only supports that for numeric expressions. For string expressions, the + will be parsed fine by the T-SQL ANTLR parser, but these will raise an error in PG.
In SQL those unary + string operators may look like they are additional concatenation operators, like: SELECT 'a' ++ 'b'. Expressions such as +++(+++@v)) are also valid syntax according to the T-SQL grammar even though they look unusual.
In this fix we remove such unary + operators, which are redundant anyway.
However we do not touch numeric constants (e.g. +123) since the +, though still redundant, may
have been included for code clarity (e.g. +123 as opposed to -123).

Signed-off-by: Rob Verschoor rcv@amazon.com

Issues Resolved
BABEL-1924 Support unary '+' operator for string expressions

tanscorpio7 committed:
fix is_member

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix is_member

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
refactor upgrade function for creating roles

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

tanscorpio7 committed:
refactor

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

Jakeowen1 committed:
Add support for Alter Function on multi-line table value functions (#2999)


This commit adds the grammar rules and logic to support alter function for multi-statement table value functions cases. The additional grammar rule is included with the other tsql_AlterFunctionStmt cases, and follows the syntax for creating multi line tvfs. The logic is the same as the other alter function cases, with the addition of handling for tabletypstmt, which represents the underlying relationship of these functions.

Issues Resolved: BABEL-5247
Signed-off-by: Jake Owen <owjco@amazon.com>

tanscorpio7 committed:
Merge branch 'server-database-roles-4-x' into BABEL-5136

tanscorpio7 commented:
Closing this PR since the target branch is moved to amazon aurora repo
New PR https://github.com/amazon-aurora/babelfish_extensions/pull/66/
